cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -739,31 +739,96 @@ private Future<MapAndStatus<ConfigMap, KafkaRebalanceStatus>> onPendingProposal(
             p.complete(buildRebalanceStatus(null, KafkaRebalanceState.Stopped, StatusUtils.validate(reconciliation, kafkaRebalance)));
         } else {
             LOGGER.infoCr(reconciliation, "Requesting a new proposal or checking the status for an already issued one");
-            requestRebalance(reconciliation, host, apiClient, kafkaRebalance, true, rebalanceOptionsBuilder,
-                    kafkaRebalance.getStatus().getSessionId())
-                .onSuccess(rebalanceMapAndStatus -> {
-                    // If the returned status has an optimization result then the rebalance proposal is ready
-                    KafkaRebalanceStatus status = rebalanceMapAndStatus.getStatus();
-                    Set<Condition> conditions = new HashSet<>();
-                    validateAnnotation(reconciliation, conditions, KafkaRebalanceState.PendingProposal, rebalanceAnnotation(kafkaRebalance), kafkaRebalance);
-                    status.addConditions(conditions);
-                    rebalanceMapAndStatus.setStatus(status);
-                    if (rebalanceMapAndStatus.getStatus().getOptimizationResult() != null &&
-                            !rebalanceMapAndStatus.getStatus().getOptimizationResult().isEmpty()) {
-                        LOGGER.infoCr(reconciliation, "Optimization proposal ready");
-                    } else {
-                        LOGGER.infoCr(reconciliation, "Waiting for optimization proposal to be ready");
-                    }
-                    p.complete(rebalanceMapAndStatus);
-                })
-                .onFailure(e -> {
-                    LOGGER.errorCr(reconciliation, "Cruise Control getting rebalance proposal failed");
-                    p.fail(e);
-                });
+            Set<Condition> conditions = StatusUtils.validate(reconciliation, kafkaRebalance);
+            validateAnnotation(reconciliation, conditions, KafkaRebalanceState.PendingProposal, rebalanceAnnotation(kafkaRebalance), kafkaRebalance);
+            String sessionId = kafkaRebalance.getStatus().getSessionId();
+            if (sessionId == null) {
+                requestRebalance(reconciliation, host, apiClient, kafkaRebalance, true, rebalanceOptionsBuilder).onSuccess(p::complete);
```

Comment:
Genuine question ... I was wondering which case you are trying to address here where there is no sessionId stored in the `KafkaRebalance` custom resource but we are in the `PendingProposal` state.

Reply:
We request a new proposal when on `onNew` state. If the response to this request is isNotEnoughDataForProposal, we return PendingProposal state without a sessionID in the rebalance status.

Also the logging on line 741 suggests that we are either requesting a new proposal or checking the status of a submitted proposal. Previously, requestRebalance() would request a new proposal if the sessionId happens to be null. So basically we would be keeping the same behaviour when using getUserTaskStatus().